### DISCUSSION TITLE
Project Vision

### DISCUSSION THREAD

**Discussion by Psypher9** [rezalas/riftshadow#231] [General]
I know that there are several things we're trying to accomplish with this project, but I want to make sure I have a good understanding of where we're going.

What are the overall goals or hopes for this project? Maybe code structure-wise, maybe feature-wise, what do you all hope to see this project become?

**Comment by sean-gilliam** [comment]
I'll defer to how the direction of the game should go to @rezalas since he's the one who started this restoration project. IIRC, though, the stated goal is to convert the project to C++ and modernize the code base to at least the C++17 standard. 

The previous devs did an amazing job with the game. However, there are some rough corners in the code that are in desperate need of smoothing over. These issues include:
- corner cases that need addressing
- better fault tolerance
- better error handling
- better logging
- better organization
- better test coverage

For me personally, I just want the game to be enjoyable for both player and dev alike.

On a side note, I can't thank you all enough for letting me be a part of this project. It's helped me dust off some serious cobwebs in my C++ skills. Plus, I love software archaeology and restoring/modernizing old code bases.

**Reply by rezalas** [reply]
I'm glad you're a part of this with us. I don't think we would be anywhere near where we are today without you, and it's always an honor to have you donate some of your time to work with us on this project.

**Comment by rezalas** [comment]
The primary goal is preservation of core gameplay throughout the migration to modern C++ (17 minimum but we're moving on up these days). Aside from the stated goal in the readme ("The primary goals for this community are to restore RS to a fully-functional state capable of operating on modern hardware with all features finished and bugs eliminated or codified as features.") I personally hope for:

- Better legibility in the codebase
- Codified documentation on how the room generation is performed, stored, and retrieved. 
  - tertiary purpose: enable future maintainers to build a graphical mapper to identify mapping issues or build new areas
- Improve operations to make the game easier for new players to enjoy (dying in the newbie area currently teleports you out of the newbie zone, which then moves you to level 20 and you can't enjoy the newbie zone anymore)
- Improve security so that characters are not so easy to hijack with modern systems.
- I'm still tossing around the idea of adding browser support so the game can be played in a Javascript environment entirely, but the idea goes along with making it more available to new audiences, as telnet is an aging standard many may not know how to access.

**Reply by rezalas** [reply]
I'm probably going to operate a server as long as I live, but the game is a part of an art that I feel would be a shame to lose. I could be wrong, but I think it's a good idea to preserve pieces of history like this.

**Comment by Psypher9** [comment]
Coming back to this conversation, how do we feel about containerizing the game server?

**Reply by sean-gilliam** [reply]
Just fyi, but I run my dev environment for this project in a Docker container. Running, testing, and debugging all work great in a container.

**Reply by Psypher9** [reply]
Cool! Is that part of the dev container that I've seen in the repo? Or something else? I'm exploring what it would be like to have PROD running in a container

**Reply by sean-gilliam** [reply]
Yeah. The `.devcontainer` folder has a Docker file and a `devcontainer.json` file. The json file is for setting up the `VS Code <-> Docker` interaction.

**Reply by Psypher9** [reply]
OHH! I see. I've not used Dev Containers before so I'm not really familiar with them.

**Reply by rezalas** [reply]
Okay, I'll try the container then! I wonder if it links up with Rodman automagically.

**Reply by sean-gilliam** [reply]
It should work with podman as well. You'd just need to change a setting in `settings.json`

from

```json
"dev.containers.dockerPath": "docker"
```

to 

```json
"dev.containers.dockerPath": "podman"
```
